File: sdk-python/src/custd/admin_tenant_storage.py
from __future__ import annotations

from typing import TypedDict, cast

from .client import AdminClient, quote_path


class TenantStorageLocation(TypedDict, total=False):
    id: str
    tenantSlug: str
    clientLocation: str
    serverAssignedPrefix: str
    status: str
    createdAt: str
    expiresAt: str


class TenantStorageCreateRequest(TypedDict, total=False):
    tenantSlug: str
    clientLocation: str


class TenantStorageListResponse(TypedDict):
    locations: list[TenantStorageLocation]


class TenantStorageClient:
    def __init__(self, admin: AdminClient) -> None:
        self._admin = admin

    def list(self) -> TenantStorageListResponse:
        return cast(TenantStorageListResponse, self._admin.request("GET", "/tenant-storage-locations"))

    def create(self, body: TenantStorageCreateRequest) -> TenantStorageLocation:
        return cast(TenantStorageLocation, self._admin.request(
            "POST", "/tenant-storage-locations", dict(body)
        ))

    def get(self, location_id: str) -> TenantStorageLocation:
        return cast(TenantStorageLocation, self._admin.request(
            "GET", f"/tenant-storage-locations/{quote_path(location_id)}"
        ))

    def revoke(self, location_id: str) -> None:
        self._admin.request(
            "DELETE", f"/tenant-storage-locations/{quote_path(location_id)}"
        )

File: sdk-js/src/admin-tenant-storage.ts
// TenantStorageClient owns tenant-scoped storage location registration.
// Locations are server-prefixed: the SDK submits clientLocation and the
// server returns a serverAssignedPrefix that the SDK must use for raw
// landing writes. Tenant is derived from the auth context; wrong-tenant
// reads collapse to an empty list indistinguishable from "no locations".

import type { RequestOptions } from "./index.js";

export type TenantStorageLocation = {
  id: string;
  tenantSlug: string;
  clientLocation: string;
  serverAssignedPrefix: string;
  status: string;
  createdAt?: string;
  expiresAt?: string;
};

export type TenantStorageListResponse = {
  locations: TenantStorageLocation[];
};

// TenantStorageCreateRequest is the body for POST /tenant-storage-locations.
// TenantSlug is server-derived; the SDK sends it for tests but the server
// is the authority for which tenant the location belongs to.
export type TenantStorageCreateRequest = {
  tenantSlug?: string;
  clientLocation: string;
};

type NonAdminRequester = <T>(method: string, path: string, body?: unknown, options?: RequestOptions) => Promise<T>;

export class TenantStorageClient {
  constructor(private readonly request: NonAdminRequester) {}

  list(options?: RequestOptions): Promise<TenantStorageListResponse> {
    return this.request("GET", "/tenant-storage-locations", undefined, options);
  }

  create(body: TenantStorageCreateRequest, options?: RequestOptions): Promise<TenantStorageLocation> {
    return this.request("POST", "/tenant-storage-locations", body, options);
  }

  get(id: string, options?: RequestOptions): Promise<TenantStorageLocation> {
    return this.request("GET", `/tenant-storage-locations/${encodeURIComponent(id)}`, undefined, options);
  }

  // Revoke removes a tenant storage location. The server is the authority for
  // whether the prefix is immediately unusable; the SDK must not assume
  // partial deletes are atomic.
  revoke(id: string, options?: RequestOptions): Promise<void> {
    return this.request("DELETE", `/tenant-storage-locations/${encodeURIComponent(id)}`, undefined, options);
  }
}
